langchain_examples/python/plugintool.py

```diff
@@ -0,0 +1,173 @@
+        )
+        base_url = base_url + f"{parsed_url.path.strip('./well-known/ai-plugin.json')}"
+
+        description = (
+            f"""Call this tool to get the OpenAPI spec (and usage guide) """
+            f"""for interacting with the {plugin.name_for_human} API. """
+            f"""Input should be either "usage_guide" or json string """
+            f"""with three keys: "path", "method", and if the method """
+            f"""is a POST request, include "data". """
+            f"""The "data" value should be a dictionary of key-value pairs """
+            f"""you want to POST to the url. """
+            f"""Please check the usage guide for POST parameters. """
+            f"""You should only call "usage_guide" ONCE! """
+            f"""This plugin is useful for {plugin.description_for_human}"""
+        )
+        open_api_spec_str = requests.get(plugin.api.url).text
+        open_api_spec = marshal_spec(open_api_spec_str)
+        openapi_schema = OpenAPISpec.from_spec_dict(open_api_spec)
+
+        spec_description = cls.generate_api_spec(openapi_schema)
+        api_spec = (
+            f"Usage Guide: {plugin.description_for_model}\n\n"
+            f"OpenAPI Spec: {spec_description}"
+        )
+
+        operations = {}
+        # TODO: Add support for other HTTP methods
+        if openapi_schema.paths:
+            for path, info in openapi_schema.paths.items():
+                if info.get:
+                    operations[(path, "get")] = APIOperation.from_openapi_spec(
+                        openapi_schema, path, "get"
+                    )
+                if info.post:
+                    operations[(path, "post")] = APIOperation.from_openapi_spec(
+                        openapi_schema, path, "post"
+                    )
+
+        print("Description: ", description)
+        return cls(
+            name=plugin.name_for_model,
+            description=description,
+            plugin=plugin,
+            api_spec=api_spec,
+            base_url=base_url,
+            operations=operations,
+            requests_wrapper=TextRequestsWrapper(),
+        )
+
+    @classmethod
+    def generate_api_spec(cls, openapi_schema: OpenAPISpec) -> str:
+        """Generate a token-optimized API spec for using the tool."""
+        operations: List[APIOperation] = []
+        if openapi_schema.paths:
+            for path, info in openapi_schema.paths.items():
+                if info.get:
+                    operations.append(
+                        APIOperation.from_openapi_spec(openapi_schema, path, "get")
+                    )
+                if info.post:
+                    operations.append(
+                        APIOperation.from_openapi_spec(openapi_schema, path, "post")
+                    )
+
+        base_str = ""
+        for op in operations:
+            op_str = f"{op.path}: ({op.method}) {op.description}\n"
+            # TODO: add support and tests for query_params
```

Comment:
Can you add tests for this before we merge this in?

I want to make sure we narrow down any possible bugs in this class before moving on